### DISCUSSION TITLE
async with message.channel.typing() only shows indicator for 10 seconds

### DISCUSSION THREAD

**Discussion by Octopirate1** [Rapptz/discord.py#5969] [Q&A] [answered]
## Summary

In the docs, `async with message.channel.typing()` says that the typing indicator is _indefinitely shown_ until the `with` context is exited. In reality, it is either 10 seconds or when the ``with`` context is exited, whichever comes first.

## Reproduction Steps

```
    if message.content.startswith('!type'):
        async with message.channel.typing():
            time.sleep(20) # or asyncio
            await message.channel.send("Time up")
```

Here, there should be a 20 second delay, during which the typing indicator in discord is active, and then a message saying "Time up". Instead, the indicator deactivates halfway through the sleep.

## Checklist

- [x] I have searched the open issues for duplicates.
- [N/A] I have shown the entire traceback, if possible.
- [x] I have removed my token from display, if visible.
- [ ] I have provided the intents that my bot is using.

## System Information

- Python v3.5.3-final
- discord.py v1.5.1-final
- aiohttp v3.6.3
- system info: Linux 4.4.0-19041-Microsoft #1-Microsoft Fri Dec 06 14:06:00 PST 2019

**Comment by Rapptz** [comment, marked as answer]
Please read the FAQ on [blocking](https://discordpy.readthedocs.io/en/latest/faq.html#what-does-blocking-mean). 

By using `time.sleep` you disallow the bot from doing anything, this includes sending the second request to continue the typing indicator.

**Reply by Octopirate1** [reply]
This is a snippet of my real code:

```
async with message.channel.typing():

            messages = await message.channel.history(limit=messagecontext).flatten()

            messagearray = ["[{}] {}#{}\n{}\n\n\n".format(message.created_at.strftime("%d-%b-%y %I:%M %p"), message.author.name, message.author.discriminator, message.content) for message in messages[1:][::-1]]
            messagearray.append("[{}] {}#{}\n".format(message.created_at.strftime("%d-%b-%y %I:%M %p"), message.author.name, message.author.discriminator))

            currenttime = "[{}] ".format(message.created_at.strftime("%d-%b-%y %I:%M %p"))

            print(messagearray)

            if "response" in seccommand or "respond" in seccommand:
                responsearray = generate_resp(''.join(messagearray)) # where the majority of time is spent

                responsesplits = [[responsesample.split("\n\n\n")] for responsesample in responsearray]
                responses = [response[0][messagecontext - 1] for response in responsesplits] # response[0] because of extra dimension
                global cutoff
                cutoff = True if responsesplits[0][0][messagecontext - 1] == responsesplits[0][0][-1] and responsesplits[0][0][messagecontext - 1][-1] != "\n" else False
                await message.channel.send(responses[0])
```

where

```
def generate_resp(text):
    response = gpt2.generate(sess,
              length=messagecontext*10,
              temperature=temp,
              prefix=text,
              return_as_list=True,
              # nsamples=2
              # batch_size=3
              )
    return response
```

`generate_resp` is a gpt-2 instance that is generating text. I've tried using ``async def generate_resp(text)`` and ``await``ing on it but that didn't seem to solve the issue at all. Should I be trying another asynchronous approach?

**Reply by Rapptz** [reply]
Since this is a blocking call then you need to turn it into an asynchronous one, presumably using [`run_in_executor`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor) on the loop.

For example:

```py
responsearray = client.loop.run_in_executor(None, generate_resp, ''.join(messagearray)) # where the majority of time is spent
```

Where `client` is replaced with the name of whatever you called your `discord.Client` or `commands.Bot` instance.

**Reply by Octopirate1** [reply]
While I believe this would work if it was any other function, TensorFlow comes with it's own problems (as it frequently does) with ``run_in_executor`` - that, however, I will solve myself. Thank you for your time.
